File: src/game/defense/tutorial/defenseTutorialAudio.ts
import {
  DEFENSE_TUTORIAL_BEATS_PER_BAR,
  DEFENSE_TUTORIAL_BPM,
  DEFENSE_TUTORIAL_LOOP_SEC,
  DEFENSE_TUTORIAL_NOTE_DURATION_SEC,
  DEFENSE_TUTORIAL_NOTE_ONSETS_SEC,
} from '@/game/defense/tutorial/defenseTutorialConstants';

const midiToFrequency = (midi: number): number => 440 * (2 ** ((midi - 69) / 12));

/** Synthesize one bar of instrument do-re-mi + rest at BPM 60 for tutorial backing. */
export const synthesizeDefenseTutorialCdeBuffer = (
  ctx: AudioContext,
  concertMidis: readonly [number, number, number],
): AudioBuffer => {
  const sampleRate = ctx.sampleRate;
  const length = Math.ceil(DEFENSE_TUTORIAL_LOOP_SEC * sampleRate);
  const buffer = ctx.createBuffer(1, length, sampleRate);
  const data = buffer.getChannelData(0);

  concertMidis.forEach((midi, index) => {
    const freq = midiToFrequency(midi);
    const startSec = DEFENSE_TUTORIAL_NOTE_ONSETS_SEC[index] ?? 0;
    const startSample = Math.floor(startSec * sampleRate);
    const durationSamples = Math.floor(DEFENSE_TUTORIAL_NOTE_DURATION_SEC * sampleRate);
    for (let i = 0; i < durationSamples; i += 1) {
      const sampleIndex = startSample + i;
      if (sampleIndex >= length) break;
      const t = i / sampleRate;
      const envelope = Math.min(1, t / 0.02) * Math.max(0, 1 - (t - 0.35) / 0.15);
      data[sampleIndex] += Math.sin(2 * Math.PI * freq * t) * envelope * 0.35;
    }
  });

  return buffer;
};

export const getDefenseTutorialTransportConfig = (): {
  bpm: number;
  beatsPerBar: number;
  loopSec: number;
} => ({
  bpm: DEFENSE_TUTORIAL_BPM,
  beatsPerBar: DEFENSE_TUTORIAL_BEATS_PER_BAR,
  loopSec: DEFENSE_TUTORIAL_LOOP_SEC,
});

File: ios/Jazzify/Defense/Tutorial/DefenseTutorialAudio.swift
import AVFoundation
import Foundation

enum DefenseTutorialAudio {
    private static func midiToFrequency(_ midi: Int) -> Double {
        440 * pow(2, Double(midi - 69) / 12)
    }

    static func synthesize(concertMidis: [Int]) -> AVAudioPCMBuffer? {
        guard concertMidis.count == 3 else { return nil }
        let format = EarTrainingAudio.preferredOutputFormat()
        let sampleRate = format.sampleRate
        let length = AVAudioFrameCount((DefenseTutorialConstants.loopSec * sampleRate).rounded(.up))
        guard let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: length) else {
            return nil
        }
        buffer.frameLength = length
        guard let channel = buffer.floatChannelData?[0] else { return nil }

        for index in 0..<3 {
            let midi = concertMidis[index]
            let freq = midiToFrequency(midi)
            let startSec = DefenseTutorialConstants.noteOnsetsSec[index]
            let startSample = Int((startSec * sampleRate).rounded(.down))
            let durationSamples = Int((DefenseTutorialConstants.noteDurationSec * sampleRate).rounded(.down))
            for offset in 0..<durationSamples {
                let sampleIndex = startSample + offset
                guard sampleIndex >= 0, sampleIndex < Int(length) else { break }
                let t = Double(offset) / sampleRate
                let attack = min(1, t / 0.02)
                let release = max(0, 1 - (t - 0.35) / 0.15)
                let envelope = attack * release
                channel[sampleIndex] += Float(sin(2 * Double.pi * freq * t) * envelope * 0.35)
            }
        }

        return buffer
    }
}
